### PR TITLE
 Adds auto-configuration feature and fixes long standing issue with validating sub-objects

### PR DESCRIPTION
**Pull request by iainmckay**
This is the first part. We'll release a new version of this then update the app to use it in a separate MR on gitlab.

### CODE REVIEW

File: src/Mapper/AutoConverterProcessor.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        );
+
+        if (! $docBlock->hasTag('var')) {
+            throw new DomainException(
+                sprintf('Property "%s" on "%s" must define a type', $property->getName(), $property->getDeclaringClass())
+            );
+        }
+
+        /* @var Var_ $var */
+        $var = $docBlock->getTagsByName('var')[0];
```
Comment (kyle-msm) [inline src/Mapper/AutoConverterProcessor.php]: I see you've got the php functional library in this project. Rather use `first`.

File: src/Mapper/ConvertingObjectMapper.php [hunk 2]
```diff
@@ -132,39 +140,66 @@ public function map($from, $to, array $metadata)
             $reflector = new \ReflectionClass($to);
 
             array_map(function (\ReflectionProperty $property) use ($from, $to, $metadata) {
-                array_map(function ($annotation) use ($property, $from, $to, $metadata) {
+                $expectedToBeAutoConfigured = false;
+
+                $convertAnnotations = array_map(function ($annotation) use ($property, $from, $to, $metadata, &$expectedToBeAutoConfigured) {
```
Comment (kyle-msm) [inline src/Mapper/ConvertingObjectMapper.php]: Can use `Functional\map` here.

File: src/Mapper/AutoConverterProcessor.php (hunk 1 above)
Comment (iainmckay) [reply src/Mapper/AutoConverterProcessor.php]: Done

File: src/Mapper/ConvertingObjectMapper.php (hunk 2 above)
Comment (iainmckay) [reply src/Mapper/ConvertingObjectMapper.php]: Done

**Review by kyle-msm — APPROVED** [review]

**Pull request merged**